### ISSUE TITLE
migrations not starting if Spring Shell is used in project

### ISSUE THREAD

**Issue by mf81bln** [KuliginStepan/mongration#18]
Hi Stepan, it's me again 🙂 

I've just stumbled upon an issue in one of our projects. If Spring Shell is integrated in the application, mongration migrations won't work, because they are triggered specifically on `ApplicationReadyEvent`. Unfortunately this event is never fired in a context where Spring Shell is used.

see: https://github.com/spring-projects/spring-shell/issues/298

I thought about different solutions and would be happy about your feedback.

---
My first (and currently used) approach is to implement my own SpringShellRunner as described in the issue linked above and start it manually:

```
public static void main(String[] args)
{
    final ConfigurableApplicationContext applicationContext = SpringApplication.run(MyApplication.class, args);
    final SpringShellRunner springShellRunner = applicationContext.getBean(SpringShellRunner.class);
    springShellRunner.start();
}
```
What I don't like with this solution:
- you have to be aware of this issue
- you have to implement your own Shell runner

---
Another solution might be, to make the event `AbstractMongration` is currently listening on configurable, for example:

MongrationProperties:
```
    /**
     * SpringApplicationEvent on which migrations should start
     */
    private Class<? extends SpringApplicationEvent> triggerEvent = ApplicationReadyEvent.class;
```

AbstractMongration:
```
    @EventListener
    public void start(SpringApplicationEvent event) {
        if (event.getClass().equals(properties.getTriggerEvent())) {
            log.info("mongration started");
            findMigrationsForExecution()
                .flatMap(tuple ->
                    withLockAcquired(Mono.defer(() -> executeMigration(tuple)))
                )
                .block();
            log.info("mongration finished its work");
        }
    }
```

This way one could just decide do override the default.

---
A third option would be to just listen to `ApplicationStartedEvent` instead of `ApplicationReadyEvent` by default but I'm not sure about possible implications... 🤔 

---
A fourth option could be to check if Spring Shell is in the classpath and then default to listen on `ApplicationStartedEvent` - but I think this adds unnecessary "dependency" to Spring Shell. And there might also be other things preventing `ApplicationReadyEvent` to be fired.

---
So I think making the trigger event configurable might be the best solution. What do you think?

Thanks in advance 🙂

**Comment by KuliginStepan** [comment]
Hi, Michel! 

If `Mongration` would listen to `SpringApplicationEvent`, we couldn't deduplicate events and `Mongration` will start several times.  
I will try to investigate listening on  `ApplicationStartedEvent` instead of `ApplicationReadyEvent` in a couple of days and also think about other possible solutions.

**Comment by KuliginStepan** [comment]
@mf81bln I will try to switch from `@EventListener` to `SmartInitializingSingleton` to start executing migrations. I assume, that these approaches are very similar, but listening on `ApplicationReadyEvent ` had one big disadvantage – then `ApplicationReadyEvent ` fired, the application already could receive requests, which is not good.

Can you try to use mongration `1.3.1-SNAPSHOT` in your environment? Snapshot maven repo – `https://oss.sonatype.org/content/repositories/snapshots`

**Comment by mf81bln** [comment]
@KuliginStepan Thanks for the quick reaction! 😃 

> listening on ApplicationReadyEvent had one big disadvantage – then ApplicationReadyEvent fired, the application already could receive requests, which is not good.

Good point, I had not thought of that at all.

I have tested your solution and it works perfectly. 👍

**Comment by KuliginStepan** [comment]
@mf81bln Released in 1.3.1
